README.md
midgit

nasmx-1.0/CHANGES.TXT
NASMX Notes:
~~~~~~~~~~~~
The project page is located at:
  http://www.asmcommunity.net/projects/nasmx/

The community forum can be found at:
  http://www.asmcommunity.net/board/index.php

Downloads and snapshots are available from:
  http://sourceforge.net/projects/nasmx/

Change Log
~~~~~~~~~~

NASMX v1.0: 20 Aug 2011, by Rob Neff

- Official release

- Updated Win32 "bin" directory with latest release
  versions of nasm tools:
    NASM v2.09.10, NDISASM v2.09.10

- Added convenience scripts for setting NASMENV

- Added Linux x64 syscall support

- Fixes to Linux x64 floating point parameters

- Added Linux x64 DEMO4

- Removed old Nasm .TXT file documentation

- Provide Nasm documentation in Adobe .PDF for Linux
  and .CHM format for Windows.
_________________________________________________________

NASMX v1.0rc2: 30 May 2011, by Rob Neff

- Added Linux x64 support

- Added Linux x64 Demos

- Fix fastcall alignment issues

- Add 32-bit Linux X window demo 6

_________________________________________________________

NASMX v1.0rc1: 27 Mar 2011, by Rob Neff
- Updated Win32 "bin" directory with latest versions
  of nasm tools:
    NASM v2.10rc4, NDISASM v2.10rc4

- Re-enable Linux 32-bit support

- Add Linux 32-bit demos

- Prevent warnings for program entry main proc
_________________________________________________________

NASMX v1.0B5: 6:15 PM 1/16/2011, by Rob Neff
- Updated Win32 "bin" directory with latest versions
  of nasm tools:
    NASM v2.10rc3, NDISASM v2.10rc3

- Support Stack Local Variable (SLV) addressing
  Ex:
    Pass contents of SLV to function:
        invoke myproc, dword [var(.x)]
    Pass address of SLV to function:
        invoke myproc, var(.x)

- Support float and double types
  Ex:
    proc myfloatproc, float_t pi

- Partial support of Win64 fastcall

- Speed up assembling Windows demos with use of master demo include
_________________________________________________________

NASMX v1.0B4: 9:25 PM 11/09/2010, by Rob Neff
- Updated Win32 "bin" directory with latest versions
  of developer tools:
    NASM v2.10rc2, NDISASM v2.10rc2,
    GoLink v0.26.14, GoRC v0.90.5
    ResEd v2.2.0.6c

- Nested structures and unions supported

- Added support for the following calling conventions:
  C, StdCall, FastCall, Pascal
  Ex:
      proto cdecl, myproc, ptrdiff_t pString

- Modified demos to reflect latest use and syntax changes
_________________________________________________________

Archive Release: 8:05 PM 1/12/2009, by Keith Kanios
- Updated Win32 "bin" directory with latest versions
  of NASM, NDISASM, GoLink, GoRC and ResEd.

- Generated new Win32 "resource.h" file based solely
  on NASMX Win32 include files.

- Removed redundant .CODE, .TEXT, .DATA and .BSS macros
  from nasmx.inc due to conflict with local labels.
_________________________________________________________

Archived Release: 9:30 PM 9/30/2007, by Keith Kanios
- Renamed project from "NASM32" to "NASMX" due to the
  inclusion of 64-bit material and cross-platform nature
  of the project.

- Revamped NASMX.INC to automatically detect object formats
  and "BITS" modes for seemless detection and use of INVOKE
  calling conventions.

- POLIB, POMAKE, PORC and POLINK have been removed in favor
  of GoLink and GoRC. GoLink supports Win64 linking.

- The Win32 DEMO batch files have been modified to reflect
  the use of GoLink and GoRC.

- The Windows installer no longer requires building the
  include and library files due to the nature and use of
  GoLink.

- The "def" and "lib" directories have been removed due to
  their obsolete nature.

- Added Win64 DEMOs 1 through 4.

- Added Linux syscall.inc in which employs the new SYSCALL
  macro that works similar to INVOKE for calling the Linux
  Kernel.

- Added Linux DEMO1 which illustrates SYSCALL usage for
  calling Linux Kernel functions.

- Changed Linux X11 DEMO1 to Destroy the Window and Close
  the Connection to X11 upon program exit.

- The "docs" directory has been deprecated and removed. All
  of the files that were in the "docs" directory have been
  placed in the root install directory.
_________________________________________________________

Archived Release: 6:30 PM 8/5/2007, by Keith Kanios
- Added X11 DEMO1, an example of a basic X11 Window.

- Added partial X11 and Linux LibC include files only
  to support the X11 DEMO1. More complete versions
  will come later.

- Compiled new Linux release package under the standard
  GZIP Tarball compression format.
_________________________________________________________

Archived Release: 7:03 AM 3/8/2007, by Keith Kanios
- Switched to using a compressed installation package.

- Added disclaimer the README.TXT file to help
  warn assembly language beginners become aware
  of the "basic" content of the packaged DEMOS.

- Removed tools from the BIN directory that were not
  vital to basic application development.

- Removed the DOS version of NASM (NASM.EXE) from the
  BIN directory, as this project has no direct intention
  of supporting the DOS environment. This also includes
  the removal of the DOS version of NDISASM (NDISASM.EXE)
  from the BIN directory as well.

- The 32-bit Windows NASM version (NASMW.EXE) has been
  renamed to simply NASM.EXE. This also includes the
  renaming of the 32-bit Windows NDISASM version
  (NDISASMW.EXE) to simply NDISASM.EXE as well.

- The Win32 and XBOX DEMO batch files have been changed
  to reflect the new layout/changes of the BIN directory.
_________________________________________________________

Archived Release: 11:51 AM 2/16/2007, by Keith Kanios
- Added "def" directory for use with export
  definition files.

- Added Win32 export definition files to "def/win32".

- Developed and added DEF2INC program for creating all
  Win32 include files as this method allows for a smaller
  package download.

- Utilizing POLIB for creating all Win32 library
  files as this method allows for a smaller package
  download.


tiny.asm
%include 'windemos.inc'


; entry point
entry entrypoint


; prototypes
proto init_intro
proto do_intro

%define WINDOWED

XRES    equ     640
YRES    equ     480

%ifdef WINDOWED
%define WIN_XRES XRES + 6
%define WIN_YRES YRES + 28
%else
%define WIN_XRES XRES
%define WIN_YRES YRES
%endif

[section .text]
proc    entrypoint, ptrdiff_t argcount, ptrdiff_t cmdline
locals none
    ; create window
    invoke  CreateWindowEx, WS_EX_APPWINDOW + WS_EX_WINDOWEDGE, szEdit, szTitle, WS_VISIBLE + WS_CAPTION + WS_CLIPSIBLINGS + WS_CLIPCHILDREN, 0, 0, WIN_XRES, WIN_YRES, 0, 0, 0, 0
    mov     ptrdiff_t [hWnd], eax
    ; get device context
    invoke  GetDC, ptrdiff_t [hWnd]
    mov     ptrdiff_t [hDc], eax
    ; initialize intro
    invoke  init_intro
    ; get initial time
    ;invoke  timeGetTime
    ;mov     dword [last_time], eax
    ; start intro loop
.intro_loop:
    ; check whether we are done
    cmp     dword [done], dword 0
    jne     .intro_end
    ; start message loop
.message_loop:
    invoke  PeekMessageA, message, 0, 0, 0, PM_REMOVE
    cmp     eax, dword 0
    je      .message_loop_end
    ; check if user pressed ESC
    invoke  GetAsyncKeyState, VK_ESCAPE
    mov     dword [done], eax
    invoke  DispatchMessage, message
    jmp     .message_loop
.message_loop_end:
    ; calculate time diff to the last frame
    ;invoke  timeGetTime
    ;mov     dword [current_time], eax
    ;mov     ebx, eax
    ;sub     eax, dword [last_time]
    ;mov     dword [diff], eax
    ;mov     dword [last_time], ebx
    ; convert integer time diff to float (scaled)
    ;fild    dword [diff]                ; load diff on TOS
    ;fld     qword [scale]               ; load scale on TOS
    ;fmul    st1                         ; scale * diff
    ;fstp    qword [fdiff]               ; store result in fdiff
    ;ffree   st0                         ; clean up TOS
    ; render frame
    invoke  do_intro
    ; show buffer
    invoke  StretchDIBits, ptrdiff_t [hDc], 0, 0, XRES, YRES, 0, 0, XRES, YRES, buffer, bmi, DIB_RGB_COLORS, SRCCOPY
    ; repeat
    jmp     .intro_loop
.intro_end:
    ; cleaning up (is this absoulutely necessary?)
    ;invoke  ReleaseDC, ptrdiff_t [hWnd], ptrdiff_t [hDc]
    ;invoke  DestroyWindow, ptrdiff_t [hWnd]
    ; exit process
    invoke  ExitProcess, NULL
endproc


proc init_intro
    locals none
endproc


proc do_intro
    locals none
endproc


[section .bss]
    hWnd:           reserve(ptrdiff_t) 1
    hDc:            reserve(ptrdiff_t) 1
;    current_time    resd 1
;    last_time       resd 1
;    diff            resd 1
;    fdiff           resq 1
    buffer          resd 307200


[section .data]
    done            dd 0
 ;   scale           dq 0.001
 ;   phase           dq 0.0
    
    szEdit:         declare(NASMX_TCHAR) NASMX_TEXT("edit"), 0x0
    szTitle:        declare(NASMX_TCHAR) NASMX_TEXT(""), 0x0
    
    NASMX_ISTRUC message, MSG
        NASMX_AT hwnd,       NULL
        NASMX_AT message,    NULL
        NASMX_AT wParam,     NULL
        NASMX_AT lParam,     NULL
        NASMX_AT time,       NULL
		NASMX_ISTRUC pt, POINT
			NASMX_AT x,          NULL
			NASMX_AT y,          NULL
		NASMX_IENDSTRUC
    NASMX_IENDSTRUC

    NASMX_ISTRUC bmpRgb, RGBQUAD
        NASMX_AT rgbBlue,           0
        NASMX_AT rgbGreen,          0
        NASMX_AT rgbRed,            0
        NASMX_AT rgbReserved,       0
    NASMX_IENDSTRUC

    NASMX_ISTRUC bmi, BITMAPINFO
        NASMX_ISTRUC bmiHeader, BITMAPINFOHEADER
            NASMX_AT biSize,            40
            NASMX_AT biWidth,           XRES
            NASMX_AT biHeight,          -YRES
            NASMX_AT biPlanes,          1
            NASMX_AT biBitCount,        32
            NASMX_AT biCompression,     BI_RGB
            NASMX_AT biSizeImage,       0
            NASMX_AT biXPelsPerMeter,   0
            NASMX_AT biYPelsPerMeter,   0
            NASMX_AT biClrUsed,         0
            NASMX_AT biClrImportant,    0
        NASMX_IENDSTRUC
        NASMX_AT bmiColors,             bmpRgb
    NASMX_IENDSTRUC

tiny_ogl.asm
global  main

extern _glOrtho@48
extern _ChangeDisplaySettingsA@8
extern _CreateWindowExA@48
extern _GetDC@4
extern _ChoosePixelFormat@8
extern _SetPixelFormat@12
extern _wglCreateContext@4
extern _wglMakeCurrent@8
extern _ShowCursor@4
extern _glViewport@16
extern _glMatrixMode@4
extern _glLoadIdentity@0
extern _glColor3ub@12
extern _glBegin@4
extern _glVertex2i@8
extern _glEnd@0
extern _SwapBuffers@4
extern _PeekMessageA@20
extern _GetAsyncKeyState@4
extern _ExitProcess@4

XRES    equ     800
YRES    equ     600
 
[section .text]
main:
    push    4                                   ; CDS_FULLSCREEN
    push    dword dms
    call    _ChangeDisplaySettingsA@8
    
    push    0
    push    0
    push    0
    push    0
    push    0
    push    0
    push    0
    push    0
    push    91000000h                           ; WS_POPUP + WS_VISIBLE + WS_MAXIMIZE
    push    0
    push    szEdit
    push    0
    call    _CreateWindowExA@48
    
    push    eax
    call    _GetDC@4
    mov     dword [hDc], eax

    push    dword pfd
    push    eax
    call    _ChoosePixelFormat@8
    
    push    dword pfd
    push    eax
    push    dword [hDc]
    call    _SetPixelFormat@12
    
    push    dword [hDc]
    call    _wglCreateContext@4

    push    eax
    push    dword [hDc]
    call    _wglMakeCurrent@8

    push    0
    call    _ShowCursor@4

    push    YRES
    push    XRES
    push    0
    push    0
    call    _glViewport@16
    
    push    0x1701
    call    _glMatrixMode@4

    call    _glLoadIdentity@0

    sub     esp, 8
    fld1
    fstp    qword [esp]
    sub     esp, 8
    fldz
    fstp    qword [esp]
    sub     esp, 8
    fld     qword [t]
    fstp    qword [esp]
    sub     esp, 8
    fldz
    fstp    qword [esp]
    sub     esp, 8
    fld     qword [r]
    fstp    qword [esp]
    sub     esp, 8
    fldz
    fstp    qword [esp]
    call    _glOrtho@48
    
    push    0x1700
    call    _glMatrixMode@4

    call    _glLoadIdentity@0

.intro_loop:
    
    push    0
    push    0
    push    255
    call    _glColor3ub@12
    
    push    0                                   ; GL_POINTS
    call    _glBegin@4
    
    push    300
    push    400
    call    _glVertex2i@8
    
    call    _glEnd@0
    
    push    dword [hDc]
    call    _SwapBuffers@4
    
    push    1                                   ; PM_REMOVE
    push    0
    push    0
    push    0
    push    0
    call    _PeekMessageA@20
    
    push    1Bh                                 ; VK_ESCAPE
    call    _GetAsyncKeyState@4
    cmp     eax, dword 0
    je      .intro_loop
    
    push    0
    call    _ExitProcess@4
    ret
    
[section .bss]
    hDc:    resd 1

[section .data]
    dms:    dd 0, 0, 0, 0, 0, 0, 0, 0, 0, 156, 1572864, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 800, 600, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
    pfd:    dd 65536, 33, 32, 0, 0, 2097152, 0, 0, 0
    r:      dq 800.0
    t:      dq 600.0
    szEdit: db 'e', 'd', 'i', 't', 0


nasmx-1.0/demos/win32/DEMO6/demo6.asm
;// DEMO6.ASM
;//
;// Copyright (C)2005-2011 The NASMX Project
;//
;// This is a fully UNICODE aware, typedefined demo that demonstrates
;// using NASMX typedef system to make your code truly portable between
;// 32 and 64-bit systems using either ASCII or UNICODE
;//
;// Contributors:
;//    Bryant Keller
;//    Rob Neff
;//
%include '..\..\windemos.inc'

[section .text]

entry    demo6

proc   dlgproc, ptrdiff_t hwnd, size_t umsg, size_t wparam, size_t lparam
locals none
    cmp      dword [argv(.umsg)], 1024
    jg       wm_default
    mov      eax, dword [argv(.umsg)]
    push     size_t [argv(.lparam)]
    push     size_t [argv(.wparam)]
    push     size_t [argv(.hwnd)]
    call     [msg_table + eax * 4]
    return

wm_default:
    xor      eax, eax
endproc

proc   Wm_DestroyProc, ptrdiff_t hwnd, size_t wparam, size_t lparam
locals none
    invoke   EndDialog, ptrdiff_t [argv(.hwnd)], 1
    return   1
endproc

proc   Wm_CommandProc, ptrdiff_t hwnd, size_t wparam, size_t lparam
uses __BX
locals none

    mov      eax, size_t [argv(.wparam)]
    jmp      .forward
.cmd_table:  dd .cmd_idgo, .cmd_idok
.forward:
    sub      eax, 200
    cmp      eax, 1
    jg       .unknown
    jmp      [.cmd_table + eax * 4]

.unknown:
    return   0

.cmd_idok:
    invoke   EndDialog, ptrdiff_t [argv(.hwnd)], 1
	return   1

.cmd_idgo:
    invoke   SendDlgItemMessage, ptrdiff_t [argv(.hwnd)], 205, WM_GETTEXTLENGTH, NULL, NULL
    cmp      eax, dword 0
    jne      .fine
    invoke   MessageBox, ptrdiff_t [argv(.hwnd)], szContent, szTitle, MB_OK | MB_ICONERROR
    return   1

.fine:
    inc      eax
    mov      ecx, eax
    push     eax
    invoke   GetProcessHeap
    mov      dword [dwHeap], eax
	pop      ecx
    invoke   HeapAlloc, eax, 0x000008, ecx 
    mov      dword [dwText], eax
    invoke   SendDlgItemMessage, ptrdiff_t [argv(.hwnd)], 205, WM_GETTEXT, eax, dwText
    invoke   SendDlgItemMessage, ptrdiff_t [argv(.hwnd)], 206, WM_SETTEXT, 0, dwText
    invoke   HeapFree, dwHeap, 0x000008, dwText
    return   1

endproc

proc   demo6, ptrdiff_t argcount, ptrdiff_t cmdline
locals none
    mov      eax, msg_table
    mov      [eax + WM_COMMAND * 4], dword Wm_CommandProc
    mov      [eax + WM_DESTROY * 4], dword Wm_DestroyProc
    invoke   DialogBoxParam, NULL, szTemplate, NULL, dlgproc, NULL
    invoke   ExitProcess, NULL
endproc

[section .bss]
    dwText:     reserve(ptrdiff_t) 1
    dwHeap:     reserve(ptrdiff_t) 1

[section .data]
    szTitle:    declare(NASMX_TCHAR) NASMX_TEXT("Demo6"), 0x0
    szContent:  declare(NASMX_TCHAR) NASMX_TEXT("Error: you must enter text into the top edit box!"), 0x0
    szTemplate: declare(NASMX_TCHAR) NASMX_TEXT("MyDialog"), 0x0
    msg_table:  times 1024*4 dd wm_default


nasmx-1.0/demos/win64/DEMO2/demo2.asm
;// DEMO2.ASM
;//
;// Copyright (C)2005-2011 The NASMX Project
;//
;// This is a fully UNICODE aware, typedefined demo that demonstrates
;// using NASMX typedef system to make your code truly portable between
;// 32 and 64-bit systems using either ASCII or UNICODE
;//
;// Contributors:
;//    Bryant Keller
;//    Rob Neff
;//

%include "..\..\windemos.inc"

DEFAULT REL

entry	demo2

[section .bss]
	fname:	resb 1025
	lname:	resb 1025
;	age:	resq 1

[section .data]
    szFirstName: declare(NASMX_TCHAR) NASMX_TEXT("First Name: "), 0
    szLastName:  declare(NASMX_TCHAR) NASMX_TEXT("Last Name: "), 0
    szAge:       declare(NASMX_TCHAR) NASMX_TEXT("Age: "), 0
	szHello:     declare(NASMX_TCHAR) 13,10,NASMX_TEXT("Hello, %s %s! What are you having today?"), 13,10,0
	szSorry:     declare(NASMX_TCHAR) 13,10,NASMX_TEXT("Sorry kid, you're only %d!"), 13,10,0
	szScanStr:   declare(NASMX_TCHAR) NASMX_TEXT("%1024s"), 0
	szScanNum:   declare(NASMX_TCHAR) NASMX_TEXT("%02d"), 0
	szPressKey:  declare(NASMX_TCHAR) 13,10,13,10,NASMX_TEXT("Press any key to exit."),13,10, 0

[section .text]

NASMX_PRAGMA CALLSTACK, 64

proc   demo2
locals
    local age, qword
endlocals

	invoke	printf, szFirstName
	invoke	scanf, szScanStr, fname
	invoke	printf, szLastName
	invoke	scanf, szScanStr, lname
	invoke	printf, szAge
	invoke	scanf, szScanNum, var(.age)
	mov	rax, qword [var(.age)]
	if	rax, >=, 21
		invoke	printf, szHello, fname, lname
	else
		invoke	printf, szSorry, qword [var(.age)]
	endif

	xor	rax, rax

endproc


nasmx-1.0/demos/win64/DEMO4/demo4.asm
;// DEMO4.ASM
;//
;// Written by Rob Neff
;// Copyright (C)2005-2011 The NASMX Project
;//
%include '..\..\windemos.inc'

DEFAULT REL

entry    demo4

[section .data]
    szStringFmt: declare(NASMX_TCHAR) 13,10,NASMX_TEXT("The area of the circle is %8.4f"), 0x0D, 0x0A, 0x0

[section .code]

NASMX_PRAGMA CALLSTACK, 32

; The following simple procedure calculates the area of a circle
; given it's radius.  It makes use of the FPU and leaves the result
; on the FPU stack for the function return value.
; Prototype:
;     float CalculateAreaOfCircle(double radius);
;
proc   CalculateAreaOfCircle, double_t radius
locals
    local pi, double_t
endlocals

    ; xmm0 contains the radius, so square the value
	movsd [argv(.radius)], xmm0
	mulsd xmm0, [argv(.radius)]

    ; assign the float value pi to local variable,
	; as there is no instruction mov mem, imm64
    mov rax,  __double(3.14159)
	mov qword [var(.pi)], rax

    ; multiply the values returning result in xmm0
    mulsd xmm0, [var(.pi)]

endproc

proc   demo4, ptrdiff_t argcount, ptrdiff_t cmdline
locals
    local area, double_t
endlocals

    ; calculate the area
    mov rcx, __double(2.0)
	mov qword [var(.area)], rcx
    invoke CalculateAreaOfCircle, rcx
;
;    Various ways to provide a double precision floating point for
;    properly prototyped procedures:
;        invoke CalculateAreaOfCircle, qword [var(.area)]
;        invoke CalculateAreaOfCircle, __double(2.0)
;        invoke CalculateAreaOfCircle, 2.0

	; vararg functions like printf are tricky regarding floating point
	; so we leave the result in xmm0 and copy it to rdx since rdx is
	; the register used for parameter 2 in win64 mode
    movq rdx, xmm0
    invoke printf, szStringFmt, rdx

endproc


nasmx-1.0/demos/win32/DEMO13/demo13.asm
;// DEMO13.ASM
;//
;// Copyright (C)2005-2011 The NASMX Project
;//
;// This is a fully UNICODE aware, type-defined demo that demonstrates
;// using NASMX typedef system to make your code truly portable between
;// 32 and 64-bit systems using either ASCII or UNICODE
;//
;// Note that you do NOT have to use the register aliases (eg: __AX)
;// as the example code below does; It simply demonstrates one way to
;// lessen the impact when porting from 32-bit to 64-bit.
;//
;// Contributors:
;//    Bryant Keller
;//    Rob Neff
;//
%include '..\..\windemos.inc'

;/////////////////////////////////////////////
;//
;// Create our own structure definition with
;// nested union and nameless struct
;//
NASMX_STRUC DEMO13_STRUC
    NASMX_RESERVE name, NASMX_TCHAR, 512
    NASMX_UNION dob
        NASMX_STRUC
            NASMX_RESERVE day,   int8_t, 1
            NASMX_RESERVE month, int8_t, 1
            NASMX_RESERVE year,  int16_t, 1
        NASMX_ENDSTRUC
        NASMX_RESERVE date, int32_t, 1
    NASMX_ENDUNION
NASMX_ENDSTRUC


;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
[section .code]
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

entry    demo13

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
proc   WndProc, ptrdiff_t hwnd, dword msg, size_t wparam, size_t lparam
locals none

.wm_create:
    cmp    dword [argv(.msg)], WM_CREATE
    jnz    .wm_command

    invoke GetClientRect, ptrdiff_t [argv(.hwnd)], rct
    invoke CreateWindowEx, NULL, szButton, szString, WS_CHILD + WS_VISIBLE,\
                           0, 0, uint32_t [rct + RECT.right], uint32_t [rct + RECT.bottom],\
                           ptrdiff_t [argv(.hwnd)], 500, ptrdiff_t [wc + WNDCLASSEX.hInstance], NULL
    jmp    .wm_default

.wm_command:
    cmp    dword [argv(.msg)], WM_COMMAND
    jnz    .wm_destroy

    cmp    dword [argv(.wparam)], 500
    jne    .wm_default

    invoke MessageBox, NULL, bday + DEMO13_STRUC.name, szTitle, MB_OK
    jmp    .exit

.wm_destroy:
    cmp    dword[argv(.msg)], WM_DESTROY
    jnz    .wm_default

    invoke PostQuitMessage, NULL

.wm_default:
    invoke   DefWindowProc, ptrdiff_t [argv(.hwnd)], dword [argv(.msg)], size_t [argv(.wparam)], size_t [argv(.lparam)]

.exit:

endproc

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
proc   WinMain, ptrdiff_t hinst, ptrdiff_t hpinst, ptrdiff_t cmdln, dword dwshow
locals none

    invoke LoadIcon, NULL, IDI_APPLICATION
    mov    __DX, __AX
    mov    __AX, ptrdiff_t [argv(.hinst)]
    mov    __BX, ptrdiff_t szClass
    mov    __CX, ptrdiff_t WndProc
    mov    ptrdiff_t [wc + WNDCLASSEX.hInstance], __AX
    mov    ptrdiff_t [wc + WNDCLASSEX.lpszClassName], __BX
    mov    ptrdiff_t [wc + WNDCLASSEX.lpfnWndProc], __CX
    mov    ptrdiff_t [wc + WNDCLASSEX.hIcon], __DX
    mov    ptrdiff_t [wc + WNDCLASSEX.hIconSm], __DX
    invoke RegisterClassEx, wc

    invoke CreateWindowEx, WS_EX_TOOLWINDOW, szClass, szTitle, WS_CAPTION + WS_SYSMENU + WS_VISIBLE,\
                           100, 120, 100, 50, NULL, NULL, ptrdiff_t [wc + WNDCLASSEX.hInstance], NULL
    mov    size_t[hWnd], __AX
    invoke ShowWindow, hWnd, dword[argv(.dwshow)]
    invoke UpdateWindow, hWnd

.msgloop:
    invoke GetMessage, message, NULL, NULL, NULL
    cmp    eax, dword 0
    je     .exit
    invoke TranslateMessage, message
    invoke DispatchMessage, message
    jmp    .msgloop

.exit:

    mov __AX, size_t[message + MSG.wParam]

endproc

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
proc   demo13, ptrdiff_t argcount, ptrdiff_t cmdline
locals none

    invoke GetModuleHandle, NULL
    mov    ptrdiff_t [hInstance], __AX
    invoke WinMain, hInstance, NULL, NULL, SW_SHOWNORMAL
    invoke ExitProcess, NULL
endproc

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
[section .bss]
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
    hInstance:  reserve(ptrdiff_t) 1
    hWnd:       reserve(ptrdiff_t) 1

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
[section .data]
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
    szButton:   declare(NASMX_TCHAR) NASMX_TEXT("BUTTON"), 0x0
    szString:   declare(NASMX_TCHAR) NASMX_TEXT("Click Me!"), 0x0
    szContent:  declare(NASMX_TCHAR) NASMX_TEXT("NASMX Demo #13"), 0x0
    szTitle:    declare(NASMX_TCHAR) NASMX_TEXT("NASMX Demo13"), 0x0
    szClass:    declare(NASMX_TCHAR) NASMX_TEXT("Demo13Class"), 0x0

    NASMX_ISTRUC bday, DEMO13_STRUC
        NASMX_AT name,   NASMX_TEXT("NASMX v1.0 Beta Birthday"),0
		NASMX_IUNION dob
			NASMX_ISTRUC
				NASMX_AT day,    8
				NASMX_AT month,  1
				NASMX_AT year,   2010
			NASMX_IENDSTRUC
		NASMX_IENDUNION
    NASMX_IENDSTRUC

    NASMX_ISTRUC wc, WNDCLASSEX
        NASMX_AT cbSize,         WNDCLASSEX_size
        NASMX_AT style,          CS_VREDRAW + CS_HREDRAW
        NASMX_AT lpfnWndProc,    NULL
        NASMX_AT cbClsExtra,     NULL
        NASMX_AT cbWndExtra,     NULL
        NASMX_AT hInstance,      NULL
        NASMX_AT hIcon,          NULL
        NASMX_AT hCursor,        NULL
        NASMX_AT hbrBackground,  COLOR_BTNFACE + 1
        NASMX_AT lpszMenuName,   NULL
        NASMX_AT lpszClassName,  NULL
        NASMX_AT hIconSm,        NULL
    NASMX_IENDSTRUC

    NASMX_ISTRUC message, MSG
        NASMX_AT hwnd,       NULL
        NASMX_AT message,    NULL
        NASMX_AT wParam,     NULL
        NASMX_AT lParam,     NULL
        NASMX_AT time,       NULL
		NASMX_ISTRUC pt, POINT
			NASMX_AT x,          NULL
			NASMX_AT y,          NULL
		NASMX_IENDSTRUC
    NASMX_IENDSTRUC

    NASMX_ISTRUC rct, RECT
        NASMX_AT left,           NULL
        NASMX_AT top,            NULL
        NASMX_AT right,          NULL
        NASMX_AT bottom,         NULL
    NASMX_IENDSTRUC
